include/bredis/AsyncConnection.hpp

```diff
@@ -0,0 +1,63 @@
+//
+//
+// Copyright (c) 2017 Ivan Baidakou (basiliscos) (the dot dmol at gmail dot com)
+//
+// Distributed under the MIT Software License
+//
+
+#pragma once
+
+#include <atomic>
+#include <initializer_list>
+#include <mutex>
+#include <queue>
+#include <sstream>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include <boost/asio.hpp>
+#include <boost/utility/string_ref.hpp>
+
+#include "Command.hpp"
+#include "Error.hpp"
+#include "Protocol.hpp"
+#include "Result.hpp"
+
+namespace bredis {
+
+template <typename AsyncStream> class AsyncConnection {
+    using protocol_type_t = typename AsyncStream::protocol_type;
+
+  public:
+    using string_t = boost::string_ref;
+    using args_container_t = std::vector<string_t>;
+
+  private:
+    AsyncStream socket_;
+
+  public:
+    template <typename... Args>
+    AsyncConnection(Args &&... args) : socket_(std::forward<Args>(args)...) {}
+
+    inline AsyncStream &next_layer() { return socket_; }
+    inline AsyncStream &&move_layer() { return std::move(socket_); }
+
+    /* asynchronous interface */
+    template <typename WriteCallback>
+    void async_write(const command_wrapper_t &command,
+                     WriteCallback write_callback);
+
+    template <typename ReadCallback, typename Buffer>
```

Comment:
Consider renaming `Buffer` to `DynamicBuffer` to make it clear what the type requirements are.